File: bit_manipulation/338.cpp
#include <iostream>
#include <vector>
using namespace std;

// https://leetcode.com/problems/counting-bits
// O(n), O(n)
class Solution {
public:
  vector<int> countBits(int n) {
    vector<int> ans(n + 1, 0);
    for (int i = 1; i <= n; i++) {
      ans[i] = ans[i >> 1] + (i & 1);
    }
    return ans;
  }
};

int main(int argc, char *argv[]) {
  int n = 5;
  cout << '[';
  for (int n: Solution().countBits(n)) {
    cout << n << ", ";
  }
  cout << ']' << endl;
  return 0;
}

File: bit_manipulation/338.py
from typing import List


# https://leetcode.com/problems/counting-bits
# O(n), O(n)
class Solution:
    def countBits(self, n: int) -> List[int]:
        dp = [0] * (n + 1)
        offset = 1
        for i in range(1, n + 1):
            if offset * 2 == i:
                offset = i
            dp[i] = 1 + dp[i - offset]

        return dp

    def bitmani(self, n: int) -> List[int]:
        ans = [0] * (n + 1)
        for i in range(1, n + 1):
            ans[i] = ans[i >> 1] + (i & 1)
        return ans


if __name__ == "__main__":
    n = 5
    print(Solution().countBits(n))
    print(Solution().bitmani(n))
